festim/hydrogen_transport_problem.py

```diff
@@ -159,39 +176,38 @@ def create_formulation(self):
         if len(self.species) > 1:
             raise NotImplementedError("Multiple species not implemented yet")
 
-        # TODO expose D_0 and E_D as parameters of a Material class
-        D_0 = fem.Constant(self.mesh.mesh, 1.9e-7)
-        E_D = fem.Constant(self.mesh.mesh, 0.2)
-
-        D = D_0 * exp(-E_D / F.k_B / self.temperature)
-
         # TODO expose dt as parameter of the model
         dt = fem.Constant(self.mesh.mesh, 1 / 20)
 
-        self.D = D  # TODO remove this
         self.dt = dt  # TODO remove this
 
+        self.formulation = 0
+
         for spe in self.species:
-            u = spe.solution
-            u_n = spe.prev_solution
-            v = spe.test_function
-
-            formulation = dot(D * grad(u), grad(v)) * self.dx
-            formulation += ((u - u_n) / dt) * v * self.dx
-
-            # add sources
-            for source in self.sources:
-                # f = Constant(my_mesh.mesh, (PETSc.ScalarType(0)))
-                if source.species == spe:
-                    formulation += source * v * self.dx
-            # add fluxes
-            # TODO implement this
-            # for bc in self.boundary_conditions:
-            #     pass
-            #     if bc.species == spe and bc.type != "dirichlet":
-            #         formulation += bc * v * self.ds
-
-        self.formulation = formulation
+            for vol in self.volume_subdomains:
+                u = spe.solution
+                u_n = spe.prev_solution
+                v = spe.test_function
```

Comment:
No need to redefine the functions for all subdomains only for each species
```suggestion
            u = spe.solution
            u_n = spe.prev_solution
            v = spe.test_function
            for vol in self.volume_subdomains:
```